pinot-segment-local/src/main/java/org/apache/pinot/segment/local/io/writer/impl/VarByteChunkSVForwardIndexWriterV4.java

```diff
@@ -0,0 +1,235 @@
+    // reserve a slot to write the data offset into
+    _output.writeInt(0);
+    _metadataSize += 4 * Integer.BYTES;
+  }
+
+  @Override
+  public void putString(String string) {
+    putBytes(string.getBytes(StandardCharsets.UTF_8));
+  }
+
+  @Override
+  public void putBytes(byte[] bytes) {
+    Preconditions.checkState(_chunkOffset < 1L << 32, "exceeded 4GB of compressed chunks");
+    int sizeRequired = Integer.BYTES + bytes.length;
+    if (_chunkBuffer.position() >= _chunkBuffer.capacity() - sizeRequired) {
+      flushChunk();
+      if (sizeRequired >= _chunkBuffer.capacity() - Integer.BYTES) {
+        writeHugeChunk(bytes);
+        return;
+      }
+    }
+    _chunkBuffer.putInt(bytes.length);
+    _chunkBuffer.put(bytes);
+    _nextDocId++;
+  }
+
+  private void writeHugeChunk(byte[] bytes) {
+    // huge values where the bytes and their length prefix don't fit in to the remainder of the buffer after the prefix
+    // for the number of documents in a regular chunk are written as a single value without metadata, and these chunks
+    // are detected by marking the MSB in the doc id offset
+    final ByteBuffer buffer;
+    if (_chunkCompressor.compressionType() == ChunkCompressionType.SNAPPY
+        || _chunkCompressor.compressionType() == ChunkCompressionType.ZSTANDARD) {
+      buffer = ByteBuffer.allocateDirect(bytes.length);
+      buffer.put(bytes);
+      buffer.flip();
+    } else {
+      // cast for JDK8 javac compatibility
+      buffer = (ByteBuffer) ByteBuffer.wrap(bytes);
+    }
+    try {
+      _nextDocId++;
+      write(buffer, true);
+    } finally {
+      CleanerUtil.cleanQuietly(buffer);
+    }
+  }
+
+  private void flushChunk() {
+    if (_nextDocId > _docIdOffset) {
+      writeChunk();
+    }
+  }
+
+  private void writeChunk() {
+    /*
+    This method translates from the current state of the buffer, assuming there are 3 values of lengths a,b, and c:
+    [-][a][a bytes][b][b bytes][c][c bytes]
+    to:
+    [3][16][a+16][a+b+16][a bytes][b bytes][c bytes]
+    [------16-bytes-----][----a+b+c bytes----------]
+     */
+    int numDocs = _nextDocId - _docIdOffset;
+    _chunkBuffer.putInt(0, numDocs);
+    // collect lengths
+    int[] valueLengths = new int[numDocs];
+    int[] offsets = new int[numDocs];
+    int offset = Integer.BYTES;
+    for (int i = 0; i < numDocs; i++) {
+      offsets[i] = offset;
+      int size = _chunkBuffer.getInt(offset);
+      valueLengths[i] = size;
+      offset += size + Integer.BYTES;
+    }
+    // now iterate backwards shifting variable length content backwards to make space for prefixes at the start
+    // this pays for itself by allowing random access to readers
+    int limit = _chunkBuffer.position();
+    int accumulatedOffset = Integer.BYTES;
+    for (int i = numDocs - 2; i >= 0; i--) {
+      ByteBuffer source = _chunkBuffer.duplicate();
+      int copyFrom = offsets[i] + Integer.BYTES;
+      source.position(offsets[i]).limit(copyFrom + valueLengths[i]);
```

Comment:
Should this be:
```suggestion
      source.position(copyFrom).limit(copyFrom + valueLengths[i]);
```

Reply:
No, this would result in an off by one (or off by 4 really) bug because of the offset for the number of docs at the start of the buffer. In fact, try it yourself and run the tests.

Reply:
Well, this was compensating for another bug, it's fixed now.